File: src/main/java/com/dungeonscribe/service/DndService.java
package com.dungeonscribe.service;

import com.dungeonscribe.dto.MagicItemDto;
import com.dungeonscribe.dto.MonsterDto;
import com.dungeonscribe.dto.Open5eResponse;
import com.dungeonscribe.dto.SpellDto;
import lombok.RequiredArgsConstructor;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestClient;

@Service
@RequiredArgsConstructor
public class DndService {

    private final RestClient open5eClient;

    public Open5eResponse<SpellDto> searchSpells(String search){
        return open5eClient.get()
                .uri("/spells/?search={search}", search)
                .retrieve()
                .body(new ParameterizedTypeReference<>() {});
    }

    public Open5eResponse<MonsterDto> searchMonsters(String search){
        return open5eClient.get()
                .uri("/monsters/?search={search}", search)
                .retrieve()
                .body(new ParameterizedTypeReference<>() {});
    }

    public Open5eResponse<MagicItemDto> searchMagicItems(String search){
        return open5eClient.get()
                .uri("/magicitems/?search={search}", search)
                .retrieve()
                .body(new ParameterizedTypeReference<>() {});
    }
}

File: react/dungeonscribe-client/src/api/dndService.ts
import api from "./axios";
import { AxiosResponse } from "axios";
import { SpellDto, MonsterDto, MagicItemDto, Open5eResponse } from "../types";

export const dndService = {
    searchSpells: (search: string) : Promise<AxiosResponse<Open5eResponse<SpellDto>>> => api.get('/dnd/spells', {params: {search}}),
    searchMonsters: (search: string) : Promise<AxiosResponse<Open5eResponse<MonsterDto>>> => api.get('/dnd/monsters', {params: {search}}),
    searchItems: (search: string) : Promise<AxiosResponse<Open5eResponse<MagicItemDto>>> => api.get('/dnd/items', {params: {search}}),
}